Sources/Basic/FileSystem.swift

```diff
@@ -276,6 +239,57 @@ private class LocalFileSystem: FileSystem {
             break
         }
     }
+
+    /// An iterator to move though contents of a directory.
+    private final class DirectoryContentsIterator: Sequence, IteratorProtocol {
+        private let dir: DirHandle?
+        private let path: AbsolutePath
+        private var entry = dirent()
+
+        private init(at path: AbsolutePath) throws {
+            guard let dir = libc.opendir(path.asString) else {
+                throw FileSystemError(errno: errno)
+            }
+            self.dir = dir
+            self.path = path
+        }
+
+        deinit { _ = libc.closedir(dir) }
+
+        public func next() -> String? {
+            while true {
+                var entryPtr: UnsafeMutablePointer<dirent>? = nil
+                if readdir_r(dir, &entry, &entryPtr) < 0 {
+                    // FIXME: Are there ever situation where we would want to
+                    // continue here?
+                    // FIXME: We're ignoring the error here, find a way to notify clients.
```

Comment:
@gribozavr has there been any direction set by Swift on how to deal with iteration + errors? The language design would seem to argue that errors are important and should be called out syntactically (i.e., `throws` and `try`), but that requires dropping out of the `Sequence` protocol.

@aciidb0mb3r we are likely to need a `DirectoryEntry` struct eventually anyway here, so what we could do is roll the error up into that (we could even roll it up into accessor functions which were `throws`, but using accessor functions doesn't feel very "Swifty"). The only other option I see is to supply our own non-`Sequence` iterator. I'm not sure which is worse...